index.html

```diff
@@ -4292,6 +4306,446 @@ <h3>Proof Formats</h3>
+<a>verifiable presentation</a>. This algorithm takes a byte sequence (bytes
+<var>inputBytes</var>) and a media type (string <var>inputMediaType</var>) as
+inputs, and produces an object (<var>verified</var>) that contains the
+following:
+        </p>
+
+        <ul>
+          <li>
+a verification status (boolean <var>status</var>)
+          </li>
+          <li>
+a verified document (object <var>document</var>)
+          </li>
+          <li>
+a media type (string <var>mediaType</var>)
+          </li>
+          <li>
+zero or more warnings (array of <a>ProblemDetails</a> <var>warnings</var>)
+          </li>
+          <li>
+zero or more errors (array of <a>ProblemDetails</a> <var>errors</var>)
+          </li>
+        </ul>
+
+        <p>
+The verification algorithm is as follows:
+        </p>
+
+        <ol class="algorithm">
+          <li>
+Set <var>verified</var> to an empty object.
+          </li>
+          <li>
+Ensure that the securing mechanism has properly protected the
+<var>conforming document</var> by performing the following steps:
+            <ol class="algorithm">
+              <li>
+Set <var>securingMechanismVerified</var> to the result of executing the
+<a href="#verify-securing-mechanism">securing mechanism verification
+algorithm</a> with <var>inputBytes</var> and <var>mediaType</var> passed as
+input to the algorithm.
+              </li>
+              <li>
+Set <var>status</var>, <var>document</var>, <var>mediaType</var>,
+<var>warnings</var>, and <var>errors</var> in <var>verified</var> to their
+respective properties in <var>securingMechanismVerified</var>.
+              </li>
+              <li>
+If <var>verified</var>.<var>status</var> is `false`, the provided
+<var>inputBytes</var> contains a document that cannot be considered to be
+secured; return <var>verified</var>, which will contain warning and error
+information related to the verification failure.
+              </li>
+            </ol>
+          </li>
+          <li>
+Ensure that the <var>verified</var>.<var>document</var> is a
+<a>conforming document</a> by performing the following steps:
+            <ol class="algorithm">
+              <li>
+Set <var>dataModelVerified</var> to the result of executing the
+<a href="#verify-data-model">data model verification algorithm</a>
+with <var>document</var> passed as input to the algorithm.
+              </li>
+              <li>
+Set <var>status</var>, <var>document</var>, <var>warnings</var>, and
+<var>errors</var> in <var>verified</var> to their respective properties in
+<var>dataModelVerified</var>.
+              </li>
+              <li>
+If <var>verified</var>.<var>status</var> is `false`, the provided
+<var>inputBytes</var> contains a non-conforming document; return
+<var>verified</var>, which will contain warning and error information
+related to the verification failure.
+              </li>
+            </ol>
+          </li>
+          <li>
+Return <var>verified</var>.
+          </li>
```

Comment:
I have still a problem with the order of steps (2) and (3) above in the case of using a JSON-LD + DI document. Per the algorithm, I have to execute the "verify securing mechanism" first, which involves the value of the `proof` property. However, _at this point_ I am not sure whether the `proof` property is used the way it must be used per specification, because checking the `proof` is step 3.4 in the data model checking algorithm which, at this point, I have not yet executed.

I believe the order of (2) and (3) depends on the securing mechanisms; the `application/*+ld_json` sequence should be the inverse of the jose-cose one.

Reply:
The conformance rules specify that you can do this step in any order as long as the end result is the same. I tried to very carefully ensure that those with different opinions could implement it however they chose.

The algorithm, as it's written now, is designed to check the digital signature first, and then check for a conforming document. If you don't have a conforming document (at least to the point that expresses the securing mechanism), then the digital signature will fail, at which point you might as well stop processing (because you can't trust the information -- even if it was a conforming document).

The other benefit for checking the digital signature first is to ensure that when you check the conforming document, that all the information you are checking was indeed protected by the digital signature.

All that said, you can do those two checks in whatever order you want.

Reply:
> All that said, you can do those two checks in whatever order you want.

Then I believe that should be made clear in the text. At the moment, the numbered list suggests required order.

Reply:
Fixed in 20aed8c6bb9fa9f9498b90ab4c2095a6c70eab82.

Reply:
Ok. I have looked at the relevant [DI specification](https://www.w3.org/TR/vc-data-integrity/#verify-proof) section, and I agree with the statement of @msporny, whereby

> If you don't have a conforming document (at least to the point that expresses the securing mechanism), then the digital signature will fail.

Indeed, the relevant steps in the DI spec do check the basic conformance statements which are relevant proof checking.

This means that the current order works and, though not fully natural when running an embedded proof, it has the advantage of working naturally when the proof is not embedded (i.e., the jose-cose case). I won't object if https://github.com/w3c/vc-data-model/commit/20aed8c6bb9fa9f9498b90ab4c2095a6c70eab82 is rolled back.

Reply:
> I don't think it's possible to reverse these steps' order and still get the same errors reported as if you hadn't reversed them

I know that some readers want there to be one path through the algorithm that everyone implements... however, there is disagreement on whether you should check the securing mechanism first and then data model conformance, or the other way around. So, I'm trying to balance those needs w/ the text (which, I admit, is not ideal).

The goal here isn't to get the exact same errors if the order is reversed, it is to get an error such that processing is halted... it doesn't really matter which error you find/report first... just that you find/report /an/ error that is in the set of errors that would all need to be resolved in order for the algorithm to complete.